### ISSUE TITLE
Jaxoplanet MCMC best practices - a rough idea of do's/dont's

### ISSUE THREAD

**Issue by TylerFair** [exoplanet-dev/jaxoplanet#267]
Currently the jaxoplanet tutoirals mostly focus on pre-MCMC with a reference to a future tutorial on best practices. I encounter some strange issues sometimes where MCMC sampling will run extremely slow sometimes and extremely fast others (which I believe is not related to the JIT component of the models), and I was looking for some tips on what those who use jax/jaxoplanet/numpyro consistently typically do for MCMC problems? I currently usually just sent up my problems as the following:

```
def get_samples(model, key, t, yerr, y, init_params):
    """
    Run MCMC to get posterior samples.
    """
    mcmc = numpyro.infer.MCMC(
        numpyro.infer.NUTS(
            model,
            regularize_mass_matrix=False, # Reccommended by tutorials
            init_strategy=numpyro.infer.init_to_value(values=init_params), # where init_params is derived from MAP estimate. 
        ),
        num_warmup=5000,
        num_samples=3000,
        jit_model_args=True 
    )
    mcmc.run(key, t, yerr, y)
    return mcmc.get_samples()
```

I am directly interested in MCMC sampling of things like the multiwavelength transit fitting tutorial, if thats any help for the kind of model set-up I am looking at (fitting hundreds of parameters simultaneously through vmap). I also know blackjax etc. is a thing, but I'm not sure if anyone's had experience in it and has found a preference for it over generic numpyro sampling?

**Comment by soichiro-hattori** [comment]
@TylerFair: thanks for bringing this up.

Have you tried running it with `dense_mass=True`? That may slow down the first part of your warmup but might make the subsequent sampling faster.

**Comment by TylerFair** [comment]
"And, not 100% sure but I think the long computational time is from sampling being slow and not a jit issue! If you diagnose your warmup you might see that your tree depths are reaching the max limit of 2^10 (it shows 1023 steps), which means NUTS is using some pretty small step sizes which might be necessary if the posterior geometry has high curvature"

>> Thats really good to know! And indeed is whats happening, a quick snapshot of some updating progress bar over time:
```
warmup:   2%|█                                                                     | 32/2000 [17:39<29:58:29, 54.83s/it, 1023 steps of size 1.58e-05. acc. prob=0.65]
warmup:   2%|█▋                                                                    | 47/2000 [30:26<29:29:36, 54.37s/it, 1023 steps of size 2.74e-05. acc. prob=0.69]
warmup:   4%|███▏                                                                  | 90/2000 [59:57<24:34:23, 46.32s/it, 1023 steps of size 1.13e-05. acc. prob=0.72]
warmup:   5%|███▍                                                               | 103/2000 [1:14:12<33:36:17, 63.77s/it, 1023 steps of size 8.19e-04. acc. prob=0.73]
warmup:   5%|███▍                                                               | 104/2000 [1:14:49<29:23:53, 55.82s/it, 1023 steps of size 1.63e-03. acc. prob=0.74]
warmup:  18%|████████████▌                                                           | 350/2000 [1:22:11<31:41,  1.15s/it, 31 steps of size 1.98e-01. acc. prob=0.78]
warmup:  32%|███████████████████████                                                 | 642/2000 [1:27:23<23:14,  1.03s/it, 31 steps of size 2.67e-01. acc. prob=0.79]
sample:  56%|███████████████████████████████████████▌                               | 1114/2000 [1:36:01<16:37,  1.13s/it, 31 steps of size 1.67e-01. acc. prob=0.88]
sample:  72%|███████████████████████████████████████████████████▍                   | 1449/2000 [1:42:38<11:24,  1.24s/it, 31 steps of size 1.67e-01. acc. prob=0.88]
sample:  95%|███████████████████████████████████████████████████████████████████▏   | 1892/2000 [1:51:30<02:04,  1.16s/it, 31 steps of size 1.67e-01. acc. prob=0.88]
sample: 100%|███████████████████████████████████████████████████████████████████████| 2000/2000 [1:53:34<00:00,  3.41s/it, 31 steps of size 1.67e-01. acc. prob=0.88]
Analysis complete!
```

I'll check out the de-centering more in detail, currently my brain can't comprehend whats going on sith the reparametrization :') 

and thank you for the update on the notebook, I am super excited to see how it can be better set-up than my current routine!

**Comment by TylerFair** [comment]
Some further insight, I found some success in reparameterizing the r-b covariance a little bit further by sampling in log_depths and a symmetrical b:
```
    _b = numpyro.sample("_b", dist.Uniform(-2.0 , 2.0 ))
    b = numpyro.deterministic('b', jnp.abs(_b))

    log_depths = numpyro.sample("log_depths", dist.Normal(jnp.log(mu_depths), 1e-2))
    depths = numpyro.deterministic('depths', jnp.exp(log_depths))
```
In my case, this improved sampling ~2x, and increased ESS for b from 300 to 1821.62 and depths from 300-500 to 1500-2000, so a massive improvement (in my understanding). This may be helpful for if you are doing more investigating @soichiro-hattori !

**Comment by TylerFair** [comment]
Going to close this issue as most of this was performed on a CPU but using a strong GPU solves multi-thousand parameter fits in minutes, so the sampling speed is no longer an issue.
